### ISSUE TITLE
Add support for auto-configuring Analytics based on service API response

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#4208]
To facilitate the onboarding experience for Google Ads, Site Kit's authentication service is now integrating with their backend, providing custom behavior for users that set up Site Kit as part of the Google Ads onboarding flow. Specifically, based on the site URL, the service will know whether the user comes from the Google Ads onboarding flow and in that case send Analytics configuration data back to Site Kit, so that the module can be auto-activated and auto-configured, without the user even having to go through the Analytics setup UI.

The authentication service will internally handle adding the OAuth scope to the list of requested scopes, so that the user does not need to go to the OAuth consent screen a second time (like they have to today when setting up Analytics after the plugin setup). The experience for users coming from the Google Ads frontend to set up Site Kit will effectively be the following: After going through the plugin setup, in addition to the regular Search Console, Analytics will already be active and connected, without the user having to do anything extra. Due to that, this feature does not encompass any UI changes, everything happens automatically.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* Listening to the `googlesitekit_authorize_user` hook, Site Kit should, if the response includes an `analytics_configuration` key, activate the Analytics module (if it's not active yet) and set the data from the associative array within as settings for the module (if it's not connected yet).
    * This bypasses both the GA4 and the existing tag check parts from the regular setup, which is acceptable for this feature.
* In case the Analytics module wasn't active before, it means that the `https://www.googleapis.com/auth/analytics.readonly` scope was stored in the "additional scopes" user setting instead of the "regular scopes" user setting. Therefore the handler that conditionally activates the Analytics module should also migrate that specific scope from the "additional scopes" user setting to the "regular scopes" user setting.

## Implementation Brief

* Note that while we only need to hook into `googlesitekit_authorize_user`, the IB splits that into two callbacks because one is Analytics-internal and needs to run regardless of whether Analytics is already active or will be activated on-the-fly, while the other one needs to activate Analytics on-the-fly, which is a core behavior and wouldn't be appropriate to trigger from within the Analytics module code.
* Implement a new public `Google\Site_Kit\Modules\Analytics::handle_token_response_data( $token_response )`. In there, if the `$token_response` includes a non-empty `analytics_configuration` key and the module is not already connected:
    * Set the `accountID`, `propertyID`, `internalWebPropertyID`, and `profileID` module settings to the values specified in the `ga_account_id`, `ua_property_id`, `ua_internal_web_property_id`, and `ua_profile_id` values within the `analytics_configuration` associative array.
* In `Google\Site_Kit\Modules\Analytics::register`, add an action for `googlesitekit_authorize_user` that calls the above `handle_token_response_data`.
* In `Google\Site_Kit\Core\Modules\Modules::register`, add an action for `googlesitekit_authorize_user` at an earlier priority than the above, e.g. 1. In there, if the `$token_response` includes a non-empty `analytics_configuration` key and Analytics is not already active:
    * Activate the Analytics module.
    * If the user's `OAuth_Client::OPTION_ADDITIONAL_AUTH_SCOPES` includes the `https://www.googleapis.com/auth/analytics.readonly` scope, remove it from there and add it to the user's `OAuth_Client::OPTION_AUTH_SCOPES` instead.
    * Call the Analytics module's `handle_token_response_data` method manually, essentially to handle the same as above on-the-fly, since the `Analytics` action hook won't be added in that case because the module wasn't active on load.

### Test Coverage

* Test coverage should be added for `Google\Site_Kit\Modules\Analytics::handle_token_response_data`.
* Specifically, other test coverage should be added to ensure correct behavior of the hooked action callbacks, for the three cases:
    * Analytics is not active (activate and set data).
    * Analytics is active, but not connected (do not activate, but set data).
    * Analytics is active and connected (do not activate and do not set data).

## QA Brief

* Test that receiving data from the proxy (only available on staging) works.

## Changelog entry

* Add support for automatically configuring Analytics with data from the service.

**Comment by felixarntz** [comment]
@eugene-manuilov @aaemnnosttv I had missed that this somehow got into EB and the current sprint already - it's actually not ready for it yet, because it depends on some changes in our authentication service that aren't done yet. The code needed here would work as is without that change even, but I'd argue it would be a bit premature (e.g. QA wouldn't be able to test that the actual feature works, it would only be possible to test that this code doesn't break anything in the existing behavior 🙃).

Also it looks like @ehoberg-g who is working on the Smart Campaigns side and has also been working on the authentication service has created a PR as well, so we now have two somewhat "conflicting" PRs. I guess there was some lack of communication here - to be fair, he is only working on this issue particularly so he didn't do the typical onboarding we do for new contributors with a more permanent role on the project.

Anyway, I should have clarified some of those things before, I guess we'll have to close one PR and combine the two.

@fhollis @eclarke1 I'm going to remove this issue from the current sprint and move it to _Stalled_ for now, just to keep it there until it's actually ready. The existing PR(s) can remain open, the code is probably already close as is, which won't change later. In other words, once this is ready to be worked on, it can go to Execution/Code Review straight away.

**Comment by felixarntz** [comment]
@eugene-manuilov This is now good to move forward with again. Let's continue with the PR #4273, I'll close #4301 as a duplicate of that. The PR mostly already looks good, let's just refresh it against latest `develop`, and there's probably small bits of feedback still.

@fhollis @eclarke1 Per the above (and since this already has an ongoing PR), we can probably add this to Sprint 66.

**Comment by FlicHollis** [comment]
@eugene-manuilov and @felixarntz this has been added to the sprint to continue execution

**Comment by felixarntz** [comment]
@wpdarren This has no QA Brief, but that's because testing is slightly complicated and it is essentially the plugin-side part that we need to test as part of https://app.asana.com/0/1199987138911229/1201364807118887/f. Let's scope this out next week.

**Comment by wpdarren** [comment]
Thanks @felixarntz - I'll look in our diaries and get something booked in so we can pick this up later this week.

**Comment by felixarntz** [comment]
@wpdarren I realize that I forgot to add this to the 1.50.0 release which will include this. I just tagged this with `QA: Eng`, and I can confirm that internally with the Smart Campaign team we've already tested this and it works as expected.

I just want to unblock this so that it doesn't need to remain open beyond the release. Obviously this is not the full QA for the epic, we'll do that separately as discussed before.

cc @fhollis @eclarke1 Per the above this actually doesn't involve anything else in Sprint 68, so feel free to remove `Rollover` and move it back to Sprint 67.
